text/0000-node-api-v2.md

```diff
@@ -0,0 +1,1077 @@
+      - [get_stempool_size](#get_stempool_size)
+      - [get_unconfirmed_transactions](#get_unconfirmed_transactions)
+      - [push_transaction](#push_transaction)
+  * [Errors](#errors)
+    + [JSON-RPC Errors](#json-rpc-errors)
+    + [API Result Errors](#api-result-errors)
+  * [Authentication](#authentication)
+    + [Wallet support](#wallet-support)
+    + [Legacy support](#legacy-support)
+    + [API only](#api-only)
+- [Drawbacks](#drawbacks)
+- [Prior art](#prior-art)
+- [Future possibilities](#future-possibilities)
+- [References](#references)
+
+# Summary
+
+[summary]: #summary
+
+Create a v2 JSON-RPC API for the Node API.
+
+# Motivation
+
+[motivation]: #motivation
+
+The current Node API (referred here as v1 API) is a REST API. This API while simple is:
+
+- Documented manually (current documentation is available [here](https://github.com/mimblewimble/grin/blob/master/doc/api/node_api.md).
+- Contains call with heterogenous args such as `?byid=xxx` and `commitment/xxx` which can be confusing and lack some uniformity.
+- Uses REST which is bound to HTTP while v2 wallet API uses JSON-RPC.
+
+The goal of this RFC is to provide a new API with:
+- Cleaner methods and errors.
+- Generated documentation directly on docs.rs.
+- Automatic testing with doc tests.
+- Stronger basis for future improvements.
+
+# Community-level explanation
+
+[community-level-explanation]: #community-level-explanation
+
+This new API will be particularly useful for wallet developer and should ultimately simplify their work on Grin. Moreover, this RFC does not introduce any breaking changes as the v1 REST API will still be around until completely deprecated.
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+## Current Endpoints with the v1 API
+
+While the endpoints are documented in details [here](https://github.com/mimblewimble/grin/blob/master/doc/api/node_api.md), here is an overview of the current REST API Endpoints of the v1 API.
+
+```JSON
+[
+  "get blocks",
+  "get headers",
+  "get chain",
+  "post chain/compact",
+  "get chain/validate",
+  "get chain/kernels/xxx?min_height=yyy&max_height=zzz",
+  "get chain/outputs/byids?id=xxx,yyy,zzz",
+  "get chain/outputs/byheight?start_height=101&end_height=200",
+  "get status",
+  "get txhashset/roots",
+  "get txhashset/lastoutputs?n=10",
+  "get txhashset/lastrangeproofs",
+  "get txhashset/lastkernels",
+  "get txhashset/outputs?start_index=1&max=100",
+  "get txhashset/merkleproof?n=1",
+  "get pool",
+  "post pool/push_tx",
+  "post peers/a.b.c.d:p/ban",
+  "post peers/a.b.c.d:p/unban",
+  "get peers/all",
+  "get peers/connected",
+  "get peers/a.b.c.d",
+  "get version"
+]
+```
+
+These endpoints can be grouped into 5 categories:
+
+- miscellaneous endpoints (which contain `status` and `version` endpoints)
+- `chain` endpoints (which also contain `blocks`, `chain` and headers` endpoints)
+- `peer` endpoints
+- `pool` endpoints
+- `txhashset` endpoints
```

Comment:
This type of analysis might make sense to be moved from the section describing existing v1 and into the section where v2 is proposed to keep all the "new thinking" in one place, like so:

> "The v2 API groups endpoints into five categories..."

Reply:
Added depreciation timeline.
Not sure I understand for the categories. Can you elaborate? I added a mention that those categories are here solely to simplify the reading of the RFC.

Reply:
👍 re the deprecation timeline. 

I agree -- grouping helps readability. But as there are no groups in v1 API today, this concept should probably be introduced in the section covering v2 of the API, rather than in the v1 section as it is in the current version of the RFC document.

Reply:
BTW by groups I mean the groups defined here: https://github.com/mimblewimble/grin/blob/master/doc/api/node_api.md. Should this be defined again for v2?

Reply:
oh! okay it's already grouped in a structure. I was confused, reading the RFC I thought you were doing this grouping for the first time here in the doc.